### PR TITLE
IPMIv2.0/RMCP+

### PR DESCRIPTION
**Pull request by IslandUsurper**
This is an initial attempt at supporting IPMI v2.0 and RMCP+. I've taken a bottom-up approach by adding and changing the functions in eipmi_auth, eipmi_encoder, and eipmi_decoder first. Still to be done are the high-level functions for eipmi and eipmi_session to handle the more complicated session routine, but I think I've done a large enough chunk to get feedback.

I have discovered that there needs to be some way to access the session state's properties from the decoder in order to decrypt and authenticate the IPMI messages. I was concerned about having to make API changes, but this would actually bring it closer to the way the encoder works, and I've already changed eipmi_auth's API anyway.

### CODE REVIEW

**Comment by schlagert** [comment]
Wow, awesome, I'll have a deeper look into it today and provide some feedback.

**Comment by schlagert** [comment]
Looks pretty good already!

Don't be too concerned about API changes, this will be a major version change anyways. Besides that the only API that is considered public is the one exposed in `eipmi.erl` along with the defines in `eipmi.hrl`. I've also seen that it will be necessary to drop support for OTP 21 which I would also be ok with.

There's no way around adding parameters to the encoder/decoder functions to be able to pass values from the session state as well as returning updated values to the session (at least from the decoder)... and of course this is perfectly fine.

Regarding whitespace and formatting: If you don't mind, I would make these consistent with the code base myself when you're done. This way you wouldn't need to think about that and can focus on the important things.

**Comment by IslandUsurper** [comment]
> Regarding whitespace and formatting
That would be helpful. I'm usually an Elixir developer, so I don't know all of the Erlang conventions and tools. If there's a quick way to set up a command that will format things for me, I'll use it, but if not I'll continue as I have.

**Comment by schlagert** [comment]
> I'll use it, but if not I'll continue as I have.

Unfortunately, we don't have a tool for formatting. Don't bother with it, I'll handle the formatting afterwards.

**Comment by IslandUsurper** [comment]
With this, I was able to activate a session over RMCP+ with the hardware I have, so it is ready for review and further refinements to the flow and functionality.

Notes on decisions made:
- I reordered the authentication types for IPMI v1.5 to prefer md5 and md2 over password and none. I expect this to be more secure by default.
- You can request which authentication type to use with `rq_auth_type`, and in fact must do this because the pre-session requests must not have an authentication code in the request. But setting `auth_type` to something besides `none` causes it to be there.
- Received messages assert that the authentication code matches the packet body.
- According to Section 13.32 Generating Additional Keying Material of the IPMI spec, I expected the constants used to be as large as the block size of the hash algorithm. But this didn't work with my hardware. However, `ipmitool` was able to make requests successfully so I found their source code to see what they did differently. [They use only 20 bytes for the constant, regardless of algorithm.](https://github.com/ipmitool/ipmitool/blob/master/src/plugins/lanplus/lanplus_crypt.c#L706) (The reasoning given is [here](https://github.com/ipmitool/ipmitool/blob/master/src/plugins/lanplus/README.lanplus).) My hardware accepted this change, so it seems like the spec doesn't always reflect reality accurately.

**Comment by schlagert** [comment]
Thanks for all this great work!

Please give me a while to look over all changes. I'll also test the stuff against our hardware (which unfortunately does not support RMCP+).

File: src/eipmi_decoder.erl [hunk 1]
```diff
@@ -183,9 +189,9 @@ session(<<?EIPMI_RESERVED:4, 6:4, E:1, A:1, P:6, I:32/little, S:32/little, L:16/
               1 -> outbound_auth_seq_nr
           end,
     {[{auth_type, rmcp_plus},
-      {encrypted, eipmi_utils:get_bool(E)},
```
Comment (schlagert) [inline src/eipmi_decoder.erl]: Sorry for messing that up 🤦

Comment (IslandUsurper) [reply src/eipmi_decoder.erl]: Oh! :sweat_smile: I had assumed I did it.

**Comment by IslandUsurper** [comment]
Since I've changed how the options to `eipmi:open/2` work a bit, here's an example of how I connect to my hardware:
```erlang
{ok, Sess} = eipmi:open(Addr, [{user, User}, {password, Pass}, {rq_auth_type, rmcp_plus}, {timeout, 5000}, {rakp_auth_type, hmac_sha256}, {integrity_type, hmac_sha256_128}, {encrypt_type, aes_cbc}]).
```
`rakp_auth_type`, `integrity_type`, and `encrypt_type` are only used if `rq_auth_type` is `rmcp_plus`. They are optional even for RMCP+, but my hardware doesn't allow any "NONE" algorithms, which are the defaults.

**Comment by schlagert** [comment]
I just ran the basic _integration tests_ against my hardware (see [eipmi_tests.erl](https://github.com/lindenbaum/eipmi/blob/master/test/eipmi_tests.erl)) and the `parallel_request_test` fails. I'm currently investigating why this is the case. The other tests are still working.

These tests are skipped by default due requiring actual IPMI-capable hardware. To execute these tests yourself you will need to set the `MCH_IP` define to a valid IP (and in your case also switch to `eipmi:open/2`).

**Comment by IslandUsurper** [comment]
I'm getting a timeout error on the `parallel_request_test`, even after I set a ridiculous value for `timeout`. Interestingly, it's throwing that error very quickly, so I think something else is happening. What error are you getting?

**Comment by schlagert** [comment]
I think I found the problem. It seems that the `rq_seq_nr` handling/management is broken. This is due to the new merge of properties in `eipmi_decoder.erl` where all properties from the session state get accidentially mangled with the `#rmcp_ipmi{}` properties.

I think it would be best to keep these two sets of properties separated, e.g. by creating `ipmi/3` instead of putting the properties into the record. Would that be possible/feasible?

**Comment by IslandUsurper** [comment]
Hm. I can see how merging in all the response properties into the state can mess things up. I don't think fixing it would make 3deb7ef obsolete, though.

On my end, it looks like only 9 responses for `get_sel_info` are coming back instead of 10.

**Comment by schlagert** [comment]
> I don't think fixing it would make https://github.com/lindenbaum/eipmi/commit/3deb7efc6aeaa1f93f578f83e1632093574bc79b obsolete, though.
Ok, that was just an assumption.

I think I will try the `ipmi/3` route and see if that fixes the unit test against my device.

**Comment by schlagert** [comment]
This version of the decoder fixes all unit tests for me: [eipmi_decoder.erl.gz](https://github.com/lindenbaum/eipmi/files/8034005/eipmi_decoder.erl.gz)

Sorry, I had to gzip the file, otherwise Github won't let me upload it as attachment.

**Comment by IslandUsurper** [comment]
That doesn't fix my test, and now I get `{error,corrupted_ipmi_message}` when using rmcp_plus. Interestingly, it doesn't happen until decoding the `set_privilege_level` response after the session is activated.

**Comment by schlagert** [comment]
Oh yes, now I see: Now the `maybe_encrypted` and `rakp` methods are missing the values in `SessionState` which are not longer contained in the `#rmcp_ipmi.properties` list.

**Comment by IslandUsurper** [comment]
After also applying #10, `get_sdr_repository` caused what looked like an infinite loop or something. My tests are failing because `parallel_request_test` only gets 9 responses back, so that might just be a hardware problem.

I also noticed that `rq_seq_nr` isn't being used right in `eipmi_session:rpc` still. It's increasing it on every retransmit, which is not correct. That number should probably be passed through the anonymous function to `handle_call`, instead of increased whenever `process_request` or `send_request` is called. I'll see if I can get that done.

**Comment by schlagert** [comment]
All tests do now succeed against my hardware :rocket: 

Great work!

**Pull request merged**